### PR TITLE
Benchmarking

### PR DESCRIPTION
**Pull request by tyurek**
This PR adds marks to many of the main functions that get executed in various protocols so that they will appear in flamegraphs. Running `cargo +nightly test --release --features flame_it --package tss-ecdsa --lib -- protocol::tests::test_run_protocol --exact --nocapture` will produce a flamegraph in stats/flame-graph.html

This flamegraph orders functions chronologically on the x axis, but if the same function is called in multiple places in the general execution flow, then it will appear broken up into various chunks, making it hard to tell how much of the overall execution is attributable to that function. 

To address this I wrote a python script (dev/flame.py) which reorganizes the flamegraph (it expects a file named flame-graph.html to be found in the same directory as it) in order to group all invocations of a given function together

### CODE REVIEW

**Comment by gatoWololo** [comment]
This PR builds on top of the changes introduced in #15. This makes it quite hard to see what is actually new and what came from #15.

Could you update this PR to only include the changes introduced in 6145c297c57308b4cacbddb11a7c77520b90cc0c?

**Comment by tyurek** [comment]
#15 changes the organization of presign quite a bit, and the main change this makes to presign is to add benchmark-related lines before the main functions that get called. The reorging is relevant to comparing benchmarks across protocols, so this PR is kind of dependent on the other PR merging anyway.

**Comment by marsella** [comment]
If you need to merge this after #15, you can change the base branch from `main` to `presign_refactor`. Then we'll only see the new changes here.

**Comment by tyurek** [comment]
Ah, thanks for the tip @marsella!

File: dev/flame.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    tail = text[-59:]
+    
+    jsonish = jsonish.replace('name:', '"name":')
+    jsonish = jsonish.replace('start:', '"start":')
+    jsonish = jsonish.replace('end:', '"end":')
+    jsonish = jsonish.replace('value:', '"value":')
+    jsonish = jsonish.replace('children:', '"children":')
+    jsonish = jsonish.replace('\n', '')
+    jsonish = jsonish.replace(',}', '}')
+    jsonish = jsonish.replace(',]', ']')
```
Comment (marsella) [inline dev/flame.py:48]: Can you leave a small inline comment about why this is happening?

**Review by marsella — COMMENTED** [review]
The code looks good, thanks for adding.

I have a few comments inline, they all boil down to "add enough docs that someone else could reasonably figure out how to use this script usefully".

**Comment by gatoWololo** [comment]
Thanks, yes. If we are planning to commit this python script into the codebase, we should add some documentation in the script on: how it is used, what the expected inputs and outputs are. As well as some comments explaining what it is doing.

File: dev/flame.py (hunk 1 above)
Comment (tyurek) [reply dev/flame.py:76]: Done. Essentially, this is to make the json module happy

File: dev/flame.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    for child_idx in range(len(blob['children'])):
+        blob['children'][child_idx]['start'] = last_end
+        blob['children'][child_idx]['end'] = last_end + blob['children'][child_idx]['value']
+        last_end = blob['children'][child_idx]['end']
+        if len(blob['children'][child_idx]['children']) > 0:
+            organize_children(blob['children'][child_idx])
+
+def parse_html():
+    #d3.select("body").datum({ children:
+    f = open("flame-graph.html", "r")
```
[the remark this answers is not in the document]
Comment (tyurek) [reply dev/flame.py:55]: I added a reference to that in file header

File: dev/flame.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    last_end = parent_start
+    for child_idx in range(len(blob['children'])):
+        blob['children'][child_idx]['start'] = last_end
+        blob['children'][child_idx]['end'] = last_end + blob['children'][child_idx]['value']
+        last_end = blob['children'][child_idx]['end']
+        if len(blob['children'][child_idx]['children']) > 0:
+            organize_children(blob['children'][child_idx])
+
+def parse_html():
+    #d3.select("body").datum({ children:
```
[the remark this answers is not in the document]
Comment (tyurek) [reply dev/flame.py]: added some clarifying context

**Review by gatoWololo — APPROVED** [review]
Looks good!

**Comment by marsella** [comment]
Thanks, docs look good. 

One last nit from me: our documentation guidelines require capital letters and end punctuation in function-level docs. This is especially important for rustdoc, but it's good practice to maintain across other languages too.

**Review by marsella — APPROVED** [review]
Thanks!

**Pull request merged**